error checking for illegal instructions

omarandlorraine opened the pull request:
I need to be able to run ill-formed code, so here is a pull-request which makes mizu-core not panic when single-stepping across an illegal opcode. I actually ran into this with the `jr fdh` instruction, which seems to jump into the offset byte, thus pointing the program counter at the illegal opcode 0xfd.

* illegal opcodes are still not handled or decoded or implemented; this just makes it possible to keep single-stepping and it returns an `Err<&'static str>` if it can't execute the instruction
* Other parts of the codebase now `.unwrap()` so the semantics are the same; this might be worth reconsidering

omarandlorraine committed:
error checking for illegal instructions

omarandlorraine commented:
the commit message should be something like "error handling for illegal opcodes" or "... for malformed instruction streams" or something

Amjad50 commented on mizu-core/src/cpu.rs:
```diff
@@ -917,16 +917,16 @@ impl Cpu {
-            Opcode::Prefix => unreachable!(),
-        };
+            Opcode::Illegal => Err("undecodable opcode"),
+            Opcode::Prefix => Err("unreachable"),
```
This is literally unreachable, so u can keep it as `unreachable!()`, as its just used during decoding

omarandlorraine replied:
OK, coming up.

Amjad50 commented on mizu-core/src/cpu.rs:
```diff
@@ -917,16 +917,16 @@ impl Cpu {
-            Opcode::Illegal => todo!(),
-            Opcode::Prefix => unreachable!(),
-        };
+            Opcode::Illegal => Err("undecodable opcode"),
```
I feel using an enum to represent the error would be better, more concrete

omarandlorraine replied:
You're right, an enum is a better idea. Coming up.

Amjad50 commented:
the idea is reasonable, but something u might need to look into is to change how we increment `pc`, because if we got an illegal instruction we do actually still increment `pc` before execution, and thus calling `next_instruction` a second time might result in valid instruction but at that point I would say its UB from pov of the CPU.

Amjad50 committed:
Fixed CI and used newer versions of actions

Amjad50 committed:
add dependabot

dependabot[bot] committed:
ci(deps): bump rustsec/audit-check from 1 to 2

Bumps [rustsec/audit-check](https://github.com/rustsec/audit-check) from 1 to 2.
- [Release notes](https://github.com/rustsec/audit-check/releases)
- [Changelog](https://github.com/rustsec/audit-check/blob/master/CHANGELOG.md)
- [Commits](https://github.com/rustsec/audit-check/compare/v1...v2)

---
updated-dependencies:
- dependency-name: rustsec/audit-check
  dependency-version: '2'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
cargo(deps): bump sfml from 0.24.0 to 0.25.1

Bumps [sfml](https://github.com/jeremyletang/rust-sfml) from 0.24.0 to 0.25.1.
- [Release notes](https://github.com/jeremyletang/rust-sfml/releases)
- [Changelog](https://github.com/jeremyletang/rust-sfml/blob/master/CHANGELOG.md)
- [Commits](https://github.com/jeremyletang/rust-sfml/compare/v0.24.0...v0.25.1)

---
updated-dependencies:
- dependency-name: sfml
  dependency-version: 0.25.1
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
cargo(deps): bump serde from 1.0.219 to 1.0.228

Bumps [serde](https://github.com/serde-rs/serde) from 1.0.219 to 1.0.228.
- [Release notes](https://github.com/serde-rs/serde/releases)
- [Commits](https://github.com/serde-rs/serde/compare/v1.0.219...v1.0.228)

---
updated-dependencies:
- dependency-name: serde
  dependency-version: 1.0.228
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
cargo(deps): bump thiserror from 2.0.12 to 2.0.17

Bumps [thiserror](https://github.com/dtolnay/thiserror) from 2.0.12 to 2.0.17.
- [Release notes](https://github.com/dtolnay/thiserror/releases)
- [Commits](https://github.com/dtolnay/thiserror/compare/2.0.12...2.0.17)

---
updated-dependencies:
- dependency-name: thiserror
  dependency-version: 2.0.17
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
cargo(deps): bump quote from 1.0.40 to 1.0.42

Bumps [quote](https://github.com/dtolnay/quote) from 1.0.40 to 1.0.42.
- [Release notes](https://github.com/dtolnay/quote/releases)
- [Commits](https://github.com/dtolnay/quote/compare/1.0.40...1.0.42)

---
updated-dependencies:
- dependency-name: quote
  dependency-version: 1.0.42
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
cargo(deps): bump clap from 4.5.40 to 4.5.53

Bumps [clap](https://github.com/clap-rs/clap) from 4.5.40 to 4.5.53.
- [Release notes](https://github.com/clap-rs/clap/releases)
- [Changelog](https://github.com/clap-rs/clap/blob/master/CHANGELOG.md)
- [Commits](https://github.com/clap-rs/clap/compare/clap_complete-v4.5.40...clap_complete-v4.5.53)

---
updated-dependencies:
- dependency-name: clap
  dependency-version: 4.5.53
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
cargo(deps): bump proc-macro2 from 1.0.95 to 1.0.103

Bumps [proc-macro2](https://github.com/dtolnay/proc-macro2) from 1.0.95 to 1.0.103.
- [Release notes](https://github.com/dtolnay/proc-macro2/releases)
- [Commits](https://github.com/dtolnay/proc-macro2/compare/1.0.95...1.0.103)

---
updated-dependencies:
- dependency-name: proc-macro2
  dependency-version: 1.0.103
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

omarandlorraine committed:
You're right that an enum is a better idea

omarandlorraine commented:
>and thus calling next_instruction a second time might result in valid instruction but at that point I would say its UB from pov of the CPU.

Do you think we need another variant in `CpuState` then? Something like `UndefinedBehavior`?

Amjad50 commented:
> >and thus calling next_instruction a second time might result in valid instruction but at that point I would say its UB from pov of the CPU.
> 
> Do you think we need another variant in `CpuState` then? Something like `UndefinedBehavior`?

No, I think when we decode an instruction, it its invalid, return error at that point and return the PC to the previous position.
I.e. don't even go to execute.

In execute function we can set it as unreachable in the match

omarandlorraine commented:
Codecov notification said the branch was behind master so I rebased it. I hope that was the right thing to do!

omarandlorraine committed:
return pc to previous value if not legal opcode

omarandlorraine commented:
Like this?

Amjad50 commented:
Yes this works

Amjad50 commented:
new version published `1.3.0`.
Thanks for the PR!
